### PR TITLE
Create y2remote libdir and clean code 

### PR DESCRIPTION
**Pull request by teclator**
## Installation Screenshots

remoteinstaller

## Screenshots in a running system

disable
allowwithsession
allowweb

### CODE REVIEW

File: src/lib/y2remote/clients/proposal.rb [hunk 1]
```diff
@@ -0,0 +1,85 @@
+#!/usr/bin/env ruby
```
Comment (imobachgs) [inline src/lib/y2remote/clients/proposal.rb]: np: this file is not supposed to be "executed", so I would remove the shebang.

File: src/lib/y2remote/clients/remote.rb [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  module Clients
+    class Remote
+      include Yast::Logger
+      include Yast::I18n
+      include Yast::UIShortcuts
+
+      def initialize
+        Yast.import "Label"
+        Yast.import "Wizard"
+        Yast.import "Report"
```
Comment (imobachgs) [inline src/lib/y2remote/clients/remote.rb]: Is this module still being used here?

File: src/lib/y2remote/clients/remote.rb [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        Yast.import "SuSEFirewall"
+
+        Yast.import "CommandLine"
+        Yast.import "RichText"
+        Yast.import "UI"
```
Comment (imobachgs) [inline src/lib/y2remote/clients/remote.rb]: It looks like CommandLine, RichtText and UI are not used here.

File: src/lib/y2remote/clients/remote.rb [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+        textdomain "network"
+      end
+
+      def remote
+        @remote = Y2Remote::Remote.instance
```
Comment (imobachgs) [inline src/lib/y2remote/clients/remote.rb]: `@remote ||= ...`

File: src/lib/y2remote/clients/remote.rb [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+      end
+
+      def run
+        log.info("----------------------------------------")
+        log.info("Remote client started")
+
+        remote.read
+        Yast::SuSEFirewall.Read
```
Comment (imobachgs) [inline src/lib/y2remote/clients/remote.rb]: How does it play with firewalld integration? (just asking...)

File: src/lib/y2remote/modes.rb [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+      include Yast::Logger
+      include Yast::I18n
+      extend Yast::I18n
+
+      class << self
+        def to_sym
+          name.split("::").last.downcase.to_sym
+        end
+
+        def installed?
```
Comment (imobachgs) [inline src/lib/y2remote/modes.rb]: I know it is trivial but, please, document.

File: src/lib/y2remote/modes.rb [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+          Yast::Package.InstalledAll(required_packages)
+        end
+
+        def required_packages
```
Comment (imobachgs) [inline src/lib/y2remote/modes.rb]: Please, document.

File: src/lib/y2remote/remote.rb [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+module Y2Remote
+  class Remote
+    include Singleton
+    include Yast::Logger
+    include Yast::I18n
+
+    XDM_SERVICE_NAME = "display-manager".freeze
+    GRAPHICAL_TARGET = "graphical".freeze
+
+    PKG_CONTAINING_FW_SERVICES = "xorg-x11-Xvnc".freeze
```
Comment (imobachgs) [inline src/lib/y2remote/remote.rb]: np: What about something like `FW_SERVICES_PACKAGE`?

File: src/lib/y2remote/remote.rb [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+
+    # Currently, all attributes (enablement of remote access)
+    # are applied on vnc1 even vnchttpd1 configuration
+
+    # [Symbol] Remote administration mode, :disabled, :xvnc or :vncmanager
```
Comment (imobachgs) [inline src/lib/y2remote/remote.rb]: I think you should add the `@return` tag (so yardoc could make a better job).

File: src/lib/y2remote/remote.rb [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+    attr_reader :modes
+
+    attr_accessor :proposed
```
Comment (imobachgs) [inline src/lib/y2remote/remote.rb]: Please, document.

File: src/lib/y2remote/remote.rb [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+    def enabled?
+      !disabled?
+    end
+
+    # Checks if remote administration is currently disallowed
+    def disabled?
+      modes.empty?
+    end
+
+    def disable!
```
Comment (imobachgs) [inline src/lib/y2remote/remote.rb]: Please, document.

File: src/lib/y2remote/remote.rb [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+      @modes = []
+    end
+
+    # It add the given mode to the list of modes to be enabled
+    #
+    # @return [Array<Symbol>] list of enable vnc modes
+    def enable_mode(mode)
+      return modes if modes.include?(mode)
+
+      @modes.delete(:vnc) if mode == :manager
```
Comment (imobachgs) [inline src/lib/y2remote/remote.rb]: np: I would prefer something like:

```ruby
if mode == :manager
  @modes.delete(:vnc)
elsif mode == :vnc
  @modes.delete(:manager)
end
```

Not a big deal, but it reads better IMHO.

File: src/lib/y2remote/remote.rb [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+    # Checks if remote administration is currently disallowed
+    def disabled?
+      modes.empty?
+    end
+
+    def disable!
+      @modes = []
+    end
+
+    # It add the given mode to the list of modes to be enabled
```
Comment (imobachgs) [inline src/lib/y2remote/remote.rb]: `It add` -> `It adds` or just `Add` (or `Adds`).

File: src/lib/y2remote/remote.rb [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+      propose!
+    end
+
+    # It propose the vnc configuration if it has not been proposed yet
+    #
+    # @return [Boolean]
+    def propose!
+      return false if proposed?
+
+      Yast::Linuxrc.vnc ? enable! : disable!
```
Comment (imobachgs) [inline src/lib/y2remote/remote.rb]: Why using bang methods? I am not against them, just asking.

File: src/lib/y2remote/remote.rb [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+
+    # Reset the proposal configuration enabling vnc if it was enabled by
+    # linuxrc
+    def reset!
+      @proposed = false
+
+      propose!
+    end
+
+    # It propose the vnc configuration if it has not been proposed yet
```
Comment (imobachgs) [inline src/lib/y2remote/remote.rb]: `It propose` -> `It proposes` or better `Propose` or `Proposes`.

File: src/lib/y2remote/remote.rb [hunk 16]
```diff
@@ -0,0 +1,233 @@
...
+      Yast::SCR.Write(
+        Yast.path(".sysconfig.displaymanager.DISPLAYMANAGER_ROOT_LOGIN_REMOTE"),
+        enabled? ? "yes" : "no"
+      )
+      Yast::SCR.Write(Yast.path(".sysconfig.displaymanager"), nil)
+
+      true
+    end
+
+    def restart_display_manager
```
Comment (imobachgs) [inline src/lib/y2remote/remote.rb]: Please, document.

File: src/lib/y2remote/remote.rb [hunk 17]
```diff
@@ -0,0 +1,233 @@
...
+      if Yast::Service.active?(XDM_SERVICE_NAME)
+        Yast::Report.Error(
```
Comment (imobachgs) [inline src/lib/y2remote/remote.rb]: np: As you report the very same error a few lines below, what about moving it to a separate method and make this line more readable?

File: src/lib/y2remote/remote.rb [hunk 18]
```diff
@@ -0,0 +1,233 @@
...
+          Yast::Message.CannotRestartService(XDM_SERVICE_NAME)
+        )
+      end
+    end
+
+    # Restarts services, reporting errors to the user
+    def restart_services
+      Yast::SystemdTarget.set_default(GRAPHICAL_TARGET) if enabled?
+
+      Y2Remote::Modes.all { |m| modes.include?(m.to_sym) ? m.restart! : m.stop! }
```
Comment (imobachgs) [inline src/lib/y2remote/remote.rb]: You could implement this logic in a class method. Something like: `restart_modes(modes)` (please, find a better name :) ).

File: src/lib/y2remote/widgets/remote.rb [hunk 19]
```diff
@@ -0,0 +1,197 @@
...
+              "session, and the session may persist even if noone is connected.</p>"
+          ),
+          5901,
+          5801
+        )
+      end
+
+    private
+
+      # Convenience method for obtain a Y2Remote::Remote instance
```
Comment (imobachgs) [inline src/lib/y2remote/widgets/remote.rb]: I would write "to obtain".

File: src/lib/y2remote/remote.rb (hunk 17 above)
Comment (teclator) [reply src/lib/y2remote/remote.rb]: Completely agree, have not touched it yet at all, I was even thinking in add a class for DisplayManager which handle this and other logic.

File: src/lib/y2remote/remote.rb (hunk 18 above)
Comment (teclator) [reply src/lib/y2remote/remote.rb]: true :)

File: src/lib/y2remote/modes.rb [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+module Y2Remote
+  class Modes
+    MODES = [VNC, Manager, Web].freeze
+
+    class << self
+      def all
+        MODES
+      end
+
+      def running_modes
```
Comment (imobachgs) [inline src/lib/y2remote/modes.rb]: please, document

File: src/lib/y2remote/modes.rb [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+# current contact information at www.suse.com.
+# ------------------------------------------------------------------------------
+
+require "yast"
+require "y2remote/modes/vnc.rb"
+require "y2remote/modes/manager.rb"
+require "y2remote/modes/web.rb"
+
+module Y2Remote
+  class Modes
```
Comment (imobachgs) [inline src/lib/y2remote/modes.rb]: Please, document

Comment (imobachgs) [inline src/lib/y2remote/modes.rb]: Why a class?

File: src/lib/y2remote/modes/base.rb [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+# current contact information at www.suse.com.
+# ------------------------------------------------------------------------------
+
+require "yast"
+
+module Y2Remote
+  # Class to handle the different remote running modes
+  class Modes
+    # Base class
+    class Base
```
Comment (imobachgs) [inline src/lib/y2remote/modes/base.rb]: I am not a big fan of nesting classes. I would change "class Modes" to "module Modes" and move `Modes` methods to `Modes::Base`.

File: src/lib/y2remote/remote.rb [hunk 23]
```diff
@@ -0,0 +1,242 @@
...
+
+      propose!
+    end
+
+    # Propose the vnc configuration if it has not been proposed yet
+    #
+    # @return [Boolean]
+    def propose!
+      return false if proposed?
+
```
Comment (imobachgs) [inline src/lib/y2remote/remote.rb]: np: TooManyBlankLinesError :)

File: test/lib/y2remote/display_manager_test.rb [hunk 24]
```diff
@@ -0,0 +1,203 @@
...
+# current contact information at www.suse.com.
+# ------------------------------------------------------------------------------
+
+require_relative "../../test_helper.rb"
+require "y2remote/remote"
+
+describe Y2Remote::DisplayManager do
+  subject { described_class.instance }
+
+  describe ".enabled?" do
```
Comment (imobachgs) [inline test/lib/y2remote/display_manager_test.rb]: It is a singleton, but this is an instance method (you should use `#enabled?`). Same for the rest of methods.

File: test/lib/y2remote/display_manager_test.rb [hunk 25]
```diff
@@ -0,0 +1,203 @@
...
+    end
+  end
+
+  describe ".remote_access?" do
+    let(:remote_access_path) { ".sysconfig.displaymanager.DISPLAYMANAGER_REMOTE_ACCESS" }
+    let(:remote_access_value) { "yes" }
+    let(:service_enabled) { true }
+
+    before do
+      allow(Yast).to receive(:path).with(remote_access_path)
```
Comment (imobachgs) [inline test/lib/y2remote/display_manager_test.rb]: np: Do you really need to mock this method? You could do something like:

```ruby
allow(Yast::SCR).to receive(:Read).with(Yast::Path.new(remote_access_path))
  .and_return(remote_access_value)
```

And get rid of the `:path` mock.

File: test/lib/y2remote/widgets/remote_test.rb [hunk 26]
```diff
@@ -0,0 +1,155 @@
...
+describe Y2Remote::Widgets do
+  let(:remote) { Y2Remote::Remote.instance }
+  before do
+    stub_const("Yast::Packages", double.as_null_object)
+  end
+
+  describe Y2Remote::Widgets::RemoteSettings do
+    include_examples "CWM::CustomWidget"
+
+    describe ".init" do
```
Comment (imobachgs) [inline test/lib/y2remote/widgets/remote_test.rb]: Instance method.

File: src/lib/y2remote/modes.rb [hunk 27]
```diff
@@ -28,20 +28,30 @@ module Y2Remote
   module Modes
     MODES = [VNC, Manager, Web].freeze
 
+    # Return a list with all the available Y2Remote::Remote::Base subclases
+    #
+    # @returm [Array<Y2Remote::Modes::Base>] list of available modes
```
Comment (imobachgs) [inline src/lib/y2remote/modes.rb]: return

File: src/lib/y2remote/modes.rb [hunk 28]
```diff
@@ -28,20 +28,30 @@ module Y2Remote
...
 
+    # Return a list with all the enabled Y2Remote::Modes::Base instances.
+    #
+    # @returm [Array<Y2Remote::Modes::Base>] list of enabled modes
     def self.running_modes
       all.select { |m| m.instance.enabled? }.map(&:instance)
     end
 
+    # It restart all the given list of Y2Remote::Modes::Base instances and stop
+    # the rest.
```
Comment (imobachgs) [inline src/lib/y2remote/modes.rb]: missing documentation for enables_modes argument. Same below.

File: src/lib/y2remote/modes/manager.rb [hunk 29]
```diff
@@ -1,25 +1,65 @@
...
+# current contact information at www.suse.com.
+# ------------------------------------------------------------------------------
+
 require "y2remote/modes/base"
 
 module Y2Remote
   module Modes
+    # This class is reponsible of vncmanager service management
     class Manager < Base
+      # <String> Service name
```
Comment (imobachgs) [inline src/lib/y2remote/modes/manager.rb]: I wolud use `[String]` (just for consistence). Even `@return [String]` if you want.

File: src/lib/y2remote/modes/socket_based.rb [hunk 30]
```diff
@@ -0,0 +1,106 @@
+require "yast"
+
+module Y2Remote
+  module Modes
+    # Common methods for handle systemd sockets
```
Comment (imobachgs) [inline src/lib/y2remote/modes/socket_based.rb]: "for handling" or "to handle"

File: src/lib/y2remote/modes/socket_based.rb [hunk 31]
```diff
@@ -0,0 +1,106 @@
...
+        true
+      end
+
+      # Convenience method to restart the systemd socket reporting an error in
+      # case of failure. It return false if the service is not installed.
+      #
+      # @return [Boolean] return false if the systemd socket is not present or
+      # not restarted; true if restarted with success
+      def restart!
+        return false unless socket
```
Comment (imobachgs) [inline src/lib/y2remote/modes/socket_based.rb]: np: Why not combine two conditions?

File: src/lib/y2remote/modes/socket_based.rb [hunk 32]
```diff
@@ -0,0 +1,106 @@
...
+        true
+      end
+
+      # Convenience method to disable the systemd socket reporting an error in
+      # case of failure. It return false if the service is not installed.
+      #
+      # @return [Boolean] return false if the systemd socket is not present or
+      # not disabled; true if disabled with success
+      def disable!
+        return false unless socket
```
Comment (imobachgs) [inline src/lib/y2remote/modes/socket_based.rb]: np: Why not combine both conditions?

**Pull request merged**